### PR TITLE
colored-terminal-output

### PR DESCRIPTION
**Pull request by abhakta-47**
add colored console output as mentioned in #103

### CODE REVIEW

**Comment by abhakta-47** [comment]
`make build` working. But, `make check` is having errors.
All the errors are related to some atomic, used by gtest. I am not sure what change is causing all the errors.

**Comment by abhakta-47** [comment]
cmake and ctest are working fine thats why I didnt see this errors.
make build also working but make check giving error regarding some atomiticity related to gtest.

After further inspection, including `private/target/stream.h` in `stumpless/target/stream.h` is causing this error.
To define stumpless_set_severity_color fucntion signature
```
void 
stumpless_set_severity_color( struct stream_target *target,
                              const enum stumpless_severity severity,
                              const char *escape_code );
```
stream_target is needed from private/stream.h

**Comment by goatshriek** [comment]
It looks like you are getting there! Changing the public-facing signature to use the generic target structure was the correct thing to do. Looks like some memory leaks are present, causing valgrind to fail: make sure you delete the severity array in the target destructor. Alternatively, since the severity array is of fixed size, consider allocating it in the struct itself instead of with dynamic memory calls.

Let me know once tests are added, and I will do a full review.

**Comment by abhakta-47** [comment]
_Atomic error is caused by including thread_safety header in cpp test files . But, `private/target/stream.h` have to be included in `stream.cpp` to use `stream_target`. `private/target/stream.h` intern includes threadsafety. Thats causing builds to fail. 
What to do here?

File: test/function/target/stream.cpp [hunk 1]
```diff
@@ -24,6 +24,7 @@
 #include <gtest/gtest.h>
 #include "test/helper/assert.hpp"
 #include "test/helper/rfc5424.hpp"
+#include "private/target/stream.h"
```
Comment (goatshriek) [inline test/function/target/stream.cpp]: @abhakta-47  This line is likely the cause for your compilation issues: private headers are not supported outside of the library compilation itself. Your tests need to work without acceess to the internals of the structures, as a result ("black box" tests).

**Comment by abhakta-47** [comment]
First of all, I was so busy with college stuff that I couldn't make time for opensource works.

To do the testing without private/ headers, my approach is like this
```c++
  TEST( SetSeverityColor, StdoutBasicArgs ){
    struct stumpless_target *target;
    const struct stumpless_error *error;
    struct stumpless_entry *basic_entry;
    struct stumpless_entry *result;
    int log_result;

    std::FILE* ofs = tmpfile();
  
    target = stumpless_open_stream_target( "stdout-test", ofs );

    stumpless_set_severity_color( target, STUMPLESS_SEVERITY_EMERG, "\033[31m" );
    
    basic_entry = stumpless_new_entry( STUMPLESS_FACILITY_USER,
                                     STUMPLESS_SEVERITY_EMERG,
                                     "stumpless-unit-test",
                                     "basic-entry",
                                     "basic test message" );
    log_result = stumpless_add_entry( target, basic_entry );
    // std::string written_msg = ofs
    char written_msg[6];
    fseek(ofs, 0, SEEK_SET);
    fgets(written_msg, 6, ofs);
    std::cout<<written_msg;
    EXPECT_STREQ( written_msg, "\033[31m" );

  }
```
i.e. instead of writing to stdout I am writing to a file and reading to see if the color escape code is appended. Currently it only checks the pre-appended escape code (for color) not checking the RESET escape code at the end, But the entry string is being changed by adding time stamp etc. So, how to dynamically match the whole string?

**Comment by abhakta-47** [comment]
```c++
TEST( SetSeverityColor, StdoutBasicArgs ){
    struct stumpless_target *target;
    const struct stumpless_error *error;
    struct stumpless_entry *basic_entry;
    struct stumpless_entry *result;
    int log_result;

    std::FILE* ofs = tmpfile();
  
    target = stumpless_open_stream_target( "stdout-test", ofs );

    stumpless_set_severity_color( target, STUMPLESS_SEVERITY_EMERG, "\033[31m" );
    
    basic_entry = stumpless_new_entry( STUMPLESS_FACILITY_USER,
                                     STUMPLESS_SEVERITY_EMERG,
                                     "stumpless-unit-test",
                                     "basic-entry",
                                     "basic test message" );
    log_result = stumpless_add_entry( target, basic_entry );
    
    char *written_msg = NULL;
    char *reset_code = NULL;
    size_t line1,line2;
    
    fseek(ofs, 0, SEEK_SET);
    
    getline( &written_msg, &line1, ofs);    
    getline( &reset_code, &line2, ofs);
    
    written_msg[5] = '\0';

    EXPECT_STREQ( written_msg, "\033[31m" );
    EXPECT_STREQ( reset_code, "\033[0m" );

  }
```
Is this approach ok ?

**Comment by goatshriek** [comment]
Yes, that approach to testing looks fine!

I took a look at your changes, and it looks like things are closing in on finished. You appear to have some heap corruption issues, which may be stemming from the fact that new stream targets do not have their `severity_colors` array initialized to NULL pointers on creation. This is likely causing the failures you are seeing in the examples. Ideally, try to add a test that fails on this case, but in any event I believe that setting all severity color strings to NULL on creation (that is, in the `new_stream_target` function) should resolve that particular issue.

Keep it up!

**Comment by goatshriek** [comment]
Checking in to see if you intend to continue work on this, per the [contribution guidelines](https://github.com/goatshriek/stumpless/blob/latest/docs/CONTRIBUTING.md#you-still-there).

**Comment by goatshriek** [comment]
Closed as stale.
